### PR TITLE
SE - Non-nullable value types: Add NonNullableValueTypeCheck

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #6105

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 1]
```diff
@@ -46,7 +46,7 @@ public RoslynSymbolicExecution(ControlFlowGraph cfg, SymbolicCheck[] checks, Can
             {
                 throw new ArgumentException("At least one check is expected", nameof(checks));
             }
-            this.checks = new(new SymbolicCheck[] { new ConstantCheck() }.Concat(checks).ToArray());
+            this.checks = new(new SymbolicCheck[] { new NotNullCheck(), new ConstantCheck() }.Concat(checks).ToArray());
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:49]: The `NotNullCheck` sounds a bit broad for all the things the engine does. Can it be `NonNullableValueTypeCheck` to keep it's scope very limited?

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Round 1 - I didn't look at the UTs yet as I can't tell what is affected by this and what is affected by the binary part easily.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 2]
```diff
@@ -30,10 +30,15 @@ internal sealed class Binary : BranchingProcessor<IBinaryOperationWrapper>
     protected override SymbolicConstraint BoolConstraintFromOperation(SymbolicContext context, IBinaryOperationWrapper operation) =>
         BinaryConstraint(operation.OperatorKind, context.State[operation.LeftOperand], context.State[operation.RightOperand]);
 
-    protected override ProgramState LearnBranchingConstraint(ProgramState state, IBinaryOperationWrapper operation, bool falseBranch) =>
-        operation.OperatorKind.IsAnyEquality()
-            ? LearnBranchingConstraint<ObjectConstraint>(state, operation, falseBranch) ?? LearnBranchingConstraint<BoolConstraint>(state, operation, falseBranch) ?? state
-            : state;
+    protected override ProgramState LearnBranchingConstraint(ProgramState state, IBinaryOperationWrapper operation, bool falseBranch)
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:33]: I got the point. Maybe it is enough for you to have a look at this commit https://github.com/SonarSource/sonar-dotnet/pull/6864/commits/ee4465effb40252d93a76041443743d25f18ce25. It shows, the fix and there is only one test affected. If you think this is enough for you to see, I would like to keep it as part of the PR, otherwise, I would revert that commit and put it in another PR.

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:33]: I'd like the master commit history to be easier to understand too.

Please revert it and cherry-pick for another PR.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs [hunk 3]
```diff
@@ -198,60 +200,63 @@ public void Binary_BoolOperands_And()
...
+                    Tag("False || True");
+                else
+                    Tag("False || True Unreachable");
+                """;
             SETestContext.CreateCS(code, ", bool arg").Validator.ValidateTagOrder(
                 "True | True",
                 "False | True",
                 "False | False",
                 "isTrue | arg True",
+                "isTrue | arg False Unreachable",
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs:257]: _No action / archeology:_ This will go away in #6879

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 4]
```diff
@@ -138,9 +138,9 @@ public void Branching_BoolExpression_LearnsBoolConstraint()
     Tag(""Value"", value);
 }";
             var validator = SETestContext.CreateCS(code).Validator;
-            validator.ValidateTag("True", x => x.HasConstraint(BoolConstraint.True).Should().BeTrue());
+            validator.ValidateTag("True", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:141]: Why this change? I'd expect it to has both, True and NotNull now.
As the test is about bool constraints, I'm fine asserting only `BoolConstraint.True` and ignoring `NotNull` here

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 5]
```diff
@@ -138,9 +138,9 @@ public void Branching_BoolExpression_LearnsBoolConstraint()
...
             validator.ValidateTag("False", x => x.HasConstraint(BoolConstraint.False).Should().BeTrue());
-            validator.ValidateTag("Value", x => x.HasConstraint(BoolConstraint.True).Should().BeTrue());    // Visited only for "true" condition
+            validator.ValidateTag("Value", x => x.HasConstraint(BoolConstraint.True).Should().BeFalse());    // Visited only for "true" condition
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:143]: _No action / archeology:_ This will go away in #6879

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 6]
```diff
@@ -653,9 +651,7 @@ public void Branching_IsNullOperation_WithIsNullOrEmpty()
         public void Branching_IsNullOperation_Equals(string expression)
         {
             var validator = CreateIfElseEndValidatorCS(expression, OperationKind.IsNull, "string");
-            validator.TagValues("If").Should().HaveCount(2)
-                .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.Null)) // ToDo: MMF-2401 This should not be here
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:657]: 🎉

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 7]
```diff
@@ -676,7 +675,14 @@ public void Invocation_LinqEnumerable_Unknown(string itemType, string expression
 var value = arg.{expression};
 Tag(""Value"", value);";
             var validator = SETestContext.CreateCS(code, $", IEnumerable<{itemType}> arg").Validator;
-            validator.ValidateTag("Value", x => x.Should().BeNull());
+            if (itemType == "int")
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:678]: Let's split the test instead.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 8]
```diff
@@ -775,7 +781,7 @@ public void Invocation_DebugAssert_LearnsNotNullForAll_AndAlso()
 
         [TestMethod]
         public void Invocation_DebugAssert_LearnsBoolConstraint_Binary() =>
-            DebugAssertValues("arg == true", "bool").Should().HaveCount(1).And.ContainSingle(x => x.HasConstraint(BoolConstraint.True));
+            DebugAssertValues("arg == true", "bool").Should().HaveCount(1).And.ContainSingle(x => x.HasConstraint(ObjectConstraint.NotNull));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:784]: This one doesn't seem right. We've lost functionality here.
I expect it to contain both. I'm fine with asserting `True` only

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 9]
```diff
@@ -964,7 +970,6 @@ public void Invocation_Equals_SplitsToBothResults_Nullable(string left, string r
             var validator = SETestContext.CreateCS(code).Validator;
             var result = validator.Symbol("result");
             validator.TagStates("End").Should().SatisfyRespectively(
-                x => x[result].HasConstraint(BoolConstraint.True).Should().BeTrue(),
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:967]: Se the `ToDo` on the `DataRow`. This whole test method should be removed. The `DataRow` should be moved to another test method

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs [hunk 10]
```diff
@@ -990,10 +995,10 @@ public void Main()
...
 private static bool Equals(object a, object b, object c) => false;";
             var validator = SETestContext.CreateCSMethod(code).Validator;
-            validator.ValidateTag("InstanceOne", x => x.Should().BeNull());
-            validator.ValidateTag("InstanceTwo", x => x.Should().BeNull());
-            validator.ValidateTag("NoArgs", x => x.Should().BeNull());
-            validator.ValidateTag("MoreArgs", x => x.Should().BeNull());
+            validator.ValidateTag("InstanceOne", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
+            validator.ValidateTag("InstanceTwo", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
+            validator.ValidateTag("NoArgs", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
+            validator.ValidateTag("MoreArgs", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:1001]: _No action:_ We need to improve the assertions :)

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs [hunk 11]
```diff
@@ -42,8 +43,9 @@ public void Loops_InstructionVisitedMaxTwice(string loop)
             var validator = SETestContext.CreateCS(code, ", int[] items", new AddConstraintOnInvocationCheck(), new PreserveTestCheck("value")).Validator;
             validator.ValidateExitReachCount(2);
             validator.TagValues("End").Should().HaveCount(2)
-                .And.ContainSingle(x => x == null)
-                .And.ContainSingle(x => x != null && x.HasConstraint(TestConstraint.First) && !x.HasConstraint(BoolConstraint.True));
+                .And.SatisfyRespectively(
+                    x => x.AllConstraints.Should().BeEquivalentTo(new SymbolicConstraint[] { ObjectConstraint.NotNull }),
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Loops.cs:47]: No action: We desperately need to bump the assertions 😆

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Nullable.cs [hunk 12]
```diff
@@ -77,9 +77,9 @@ public void Nullable_HasValue_ReadsBoolConstraintFromObjectConstraint()
             Tag("SymbolAfterNull", arg);
             """;
         var validator = SETestContext.CreateCS(code, ", bool? arg").Validator;
-        validator.ValidateTag("HasValueUnknown", x => x.Should().BeNull());
+        validator.ValidateTag("HasValueUnknown", x => x.HasConstraint<BoolConstraint>().Should().BeFalse());
         validator.ValidateTag("SymbolUnknown", x => x.Should().BeNull());
-        validator.ValidateTag("HasValueAfterTrue", x => x.Should().BeNull());   // ToDo: Should be x.HasConstraint(BoolConstraint.True).Should().BeTrue()); after we build NotNull for bool literal.
+        validator.ValidateTag("HasValueAfterTrue", x => x.HasConstraint<BoolConstraint>().Should().BeFalse());   // ToDo: Should be x.HasConstraint(BoolConstraint.True).Should().BeTrue()); after we build NotNull for bool literal.
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Nullable.cs:82]: This is suspicious. I was expecting the `ToDo` go away and have different assertion here

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 13]
```diff
@@ -809,19 +809,19 @@ public void TypeOf_SetsNotNull()
...
-        [DataRow("bool", "true", ConstraintKind.BoolFalse)]
-        [DataRow("bool", "false", ConstraintKind.BoolTrue)]
-        [DataRow("bool?", "default", ConstraintKind.ObjectNull)]
-        [DataRow("bool?", "null", ConstraintKind.ObjectNull)]
-        public void Unary_Not_SupportsBoolAndNull(string type, string defaultValue, ConstraintKind expectedConstraint)
+        [DataRow("bool", "true", new[] { ConstraintKind.ObjectNotNull, ConstraintKind.BoolFalse })]
+        [DataRow("bool", "false", new[] { ConstraintKind.ObjectNotNull, ConstraintKind.BoolTrue })]
+        [DataRow("bool?", "default", new[] { ConstraintKind.ObjectNull })]
+        [DataRow("bool?", "null", new[] { ConstraintKind.ObjectNull })]
+        public void Unary_Not_SupportsBoolAndNull(string type, string defaultValue, ConstraintKind[] expectedConstraints)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:816]: Like this, we can remove the `new[] {` part above
```suggestion
        public void Unary_Not_SupportsBoolAndNull(string type, string defaultValue, params ConstraintKind[] expectedConstraints)
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
There are two scenarios that are concerning

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/NotNullCheck.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn.Checks
+{
+    internal class NotNullCheck : SymbolicCheck
+    {
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/NotNullCheck.cs]: I added to my working list in #6105

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Nullable.cs (hunk 12 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Nullable.cs:82]: Added as task to review later

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM
Please move the one DataRow before merging.
Rebase will be needed in any case - you can consider using the new assertions

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs (hunk 9 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs:967]: Ah. I understood you wrong. I moved the test case in 48f730e23fe6406c8e19287b22821d1a0d448e1f. Please have a second look.
